File: apps/api/src/mate/api/datasets/adapters.py
"""Server-side ``kind -> DatasetEnvelope`` adapters + module-dataset resolution.

Python port of ``apps/web/lib/visualizations/adapters.ts``. Resolves a module
dataset by invoking its route handler through the loader (reusing the module's
result cache + the per-request ephemeral filter) and normalizes the response
into the canonical :class:`DatasetEnvelope`.
"""

from __future__ import annotations

import re
from typing import Any

from mate.api.datasets.envelope import (
    ColumnSpec,
    ColumnType,
    DatasetEnvelope,
    DatasetMeta,
    DatasetSchema,
    FieldRole,
    GraphData,
    GraphEdge,
    GraphNode,
    KpiData,
    KpiFormat,
    KpiMeasure,
    TreeData,
    TreeNode,
)

_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}([T ]\d{2}:\d{2})?")
_NUMERIC: frozenset[ColumnType] = frozenset({"number", "integer", "duration"})


def _prettify(key: str) -> str:
    return re.sub(r"[_-]+", " ", key).strip().title()


def _infer_type(value: Any) -> ColumnType:
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, int):
        return "integer"
    if isinstance(value, float):
        return "number"
    if isinstance(value, str):
        if _DATE_RE.match(value):
            return "datetime"
        try:
            f = float(value)
            return "integer" if f.is_integer() else "number"
        except (TypeError, ValueError):
            return "string"
    return "string"


def _role_for(col_type: ColumnType) -> FieldRole:
    if col_type in _NUMERIC:
        return "measure"
    if col_type == "datetime":
        return "time"
    return "dimension"


def table_from_rows(rows: list[dict[str, Any]]) -> tuple[list[ColumnSpec], list[dict[str, Any]]]:
    keys: list[str] = []
    seen: set[str] = set()
    for r in rows:
        for k in r:
            if k not in seen:
                seen.add(k)
                keys.append(k)
    columns: list[ColumnSpec] = []
    for k in keys:
        sample = next((r[k] for r in rows if r.get(k) is not None), None)
        col_type = _infer_type(sample)
        columns.append(
            ColumnSpec(id=k, label=_prettify(k), type=col_type, role=_role_for(col_type))
        )
    return columns, rows


def _kpi_format(key: str, value: float) -> KpiFormat | None:
    k = key.lower()
    if re.search(r"(fitness|precision|perc|percent|share|ratio|rate)", k) and 0 <= value <= 1:
        return "percent"
    if re.search(r"(^n[_.]|count|total|num)", k) and float(value).is_integer():
        return "integer"
    return "integer" if float(value).is_integer() else "number"


def _measures_from(obj: dict[str, Any], prefix: str = "") -> list[KpiMeasure]:
    out: list[KpiMeasure] = []
    for k, v in obj.items():
        key = f"{prefix}.{k}" if prefix else k
        if isinstance(v, bool):
            continue
        if isinstance(v, (int, float)):
            out.append(
                KpiMeasure(key=key, label=_prettify(k), value=float(v), format=_kpi_format(k, v))
            )
        elif isinstance(v, dict):
            out.extend(_measures_from(v, key))
    return out


def _graph_from_dfg(raw: dict[str, Any]) -> GraphData:
    activities = raw.get("activities") or []
    edges = raw.get("edges") or []
    nodes = [
        GraphNode(
            id=str(a.get("id")),
            label=str(a.get("label", a.get("id"))),
            value=a.get("frequency") if isinstance(a.get("frequency"), (int, float)) else None,
            kind="activity",
        )
        for a in activities
    ]
    g_edges = [
        GraphEdge(
            id=str(e.get("id", f"e{i}")),
            source=str(e.get("source")),
            target=str(e.get("target")),
            value=e.get("frequency") if isinstance(e.get("frequency"), (int, float)) else None,
            performanceSeconds=e.get("performance_seconds")
            if isinstance(e.get("performance_seconds"), (int, float))
            else None,
        )
        for i, e in enumerate(edges)
    ]
    return GraphData(
        directed=True,
        nodes=nodes,
        edges=g_edges,
        start=list((raw.get("start_activities") or {}).keys()),
        end=list((raw.get("end_activities") or {}).keys()),
    )


def _graph_from_petri(raw: dict[str, Any]) -> GraphData:
    places = raw.get("places") or []
    transitions = raw.get("transitions") or []
    arcs = raw.get("arcs") or []
    nodes = [
        GraphNode(id=str(p.get("id")), label=str(p.get("label", "")), kind="place") for p in places
    ]
    nodes += [
        GraphNode(
            id=str(t.get("id")),
            label="" if t.get("is_invisible") else str(t.get("label", t.get("name", ""))),
            kind="transition",
        )
        for t in transitions
    ]
    edges = [
        GraphEdge(
            id=str(a.get("id", f"a{i}")),
            source=str(a.get("source")),
            target=str(a.get("target")),
            value=a.get("weight") if isinstance(a.get("weight"), (int, float)) else None,
        )
        for i, a in enumerate(arcs)
    ]
    return GraphData(directed=True, nodes=nodes, edges=edges)


def _tree_from_process_tree(raw: dict[str, Any]) -> TreeData:
    counter = [0]

    def walk(node: dict[str, Any]) -> TreeNode:
        children = node.get("children") or []
        label = node.get("label") or node.get("operator") or "·"
        counter[0] += 1
        return TreeNode(
            id=str(node.get("id", f"n{counter[0]}")),
            label=str(label),
            children=[walk(c) for c in children],
        )

    root = raw.get("root")
    return TreeData(root=walk(root) if isinstance(root, dict) else TreeNode(id="root", label="·"))


def _tree_from_prefix(raw: dict[str, Any]) -> TreeData:
    flat = raw.get("nodes") or []
    by_id: dict[str, TreeNode] = {
        str(n.get("id")): TreeNode(
            id=str(n.get("id")),
            label=str(n.get("label") or "start"),
            value=n.get("frequency") if isinstance(n.get("frequency"), (int, float)) else None,
        )
        for n in flat
    }
    root: TreeNode | None = None
    for n in flat:
        node = by_id[str(n.get("id"))]
        parent = by_id.get(str(n.get("parent"))) if n.get("parent") is not None else None
        if parent is not None:
            parent.children.append(node)
        elif root is None:
            root = node
    return TreeData(root=root or TreeNode(id="root", label="start"))


def adapt(shape: str, raw: Any) -> DatasetEnvelope:
    """Normalize a module route response into the declared ``shape``."""
    obj: dict[str, Any] = raw if isinstance(raw, dict) else {}
    kind = obj.get("kind") if isinstance(obj.get("kind"), str) else ""

    if shape == "graph":
        g = _graph_from_petri(obj) if kind == "petri_net" else _graph_from_dfg(obj)
        return DatasetEnvelope(shape="graph", data=g, meta=DatasetMeta(sourceKind=kind))

    if shape == "tree":
        t = _tree_from_prefix(obj) if kind == "prefix_tree" else _tree_from_process_tree(obj)
        return DatasetEnvelope(shape="tree", data=t, meta=DatasetMeta(sourceKind=kind))

    if shape == "kpi":
        items: list[KpiMeasure] = []
        note: str | None = None
        if kind == "conformance":
            if obj.get("ran") is False or not isinstance(obj.get("kpis"), dict):
                note = "No conformance run yet - run conformance on this log first."
            else:
                items = _measures_from(obj["kpis"])
        elif isinstance(obj.get("basic"), dict) or isinstance(obj.get("enriched"), dict):
            items = _measures_from(obj.get("basic") or {})
            if isinstance(obj.get("enriched"), dict):
                items += _measures_from(obj["enriched"], "enriched")
        else:
            items = _measures_from(obj)
        return DatasetEnvelope(
            shape="kpi", data=KpiData(items=items), meta=DatasetMeta(sourceKind=kind, note=note)
        )

    if shape == "blob":
        value = obj.get("xml") if isinstance(obj.get("xml"), str) else obj.get("value") or ""
        media = "bpmn-xml" if kind == "bpmn" else "text"
        from mate.api.datasets.envelope import BlobData

        return DatasetEnvelope(
            shape="blob",
            data=BlobData(media=media, value=str(value)),
            meta=DatasetMeta(sourceKind=kind),
        )

    # table (default)
    rows: list[dict[str, Any]] = []
    note = None
    if kind == "conformance":
        rows = obj.get("per_activity") or []
        if not rows and obj.get("ran") is False:
            note = "No conformance run yet."
    elif isinstance(obj.get("drifts"), list):
        rows = obj["drifts"]
    else:
        for v in obj.values():
            if isinstance(v, list) and v and isinstance(v[0], dict):
                rows = v
                break
    columns, rows = table_from_rows(rows)
    from mate.api.datasets.envelope import TableData

    return DatasetEnvelope(
        shape="table",
        schema=DatasetSchema(columns=columns),
        data=TableData(columns=columns, rows=rows),
        meta=DatasetMeta(
            sourceKind=kind, rowCount=len(rows), note=None if rows else (note or "No rows.")
        ),
    )


async def resolve_dataset(
    loader: Any,
    module_id: str,
    dataset_id: str,
    log_id: str,
    user_id: str,
    *,
    filter_override: list[dict[str, Any]] | None = None,
    params: dict[str, Any] | None = None,
    restrict_event_log: bool = False,
) -> DatasetEnvelope:
    """Resolve a module dataset to a canonical envelope by invoking its route
    handler through the loader (reuses the module's result cache + filtering).
    ``restrict_event_log=True`` walls off raw event rows (AI/MCP callers)."""
    loaded = loader.loaded.get(module_id)
    if loaded is None:
        raise ValueError(f"Module {module_id!r} is not loaded.")
    entry = next((d for d in loaded.manifest.datasets if d.id == dataset_id), None)
    if entry is None:
        raise ValueError(f"Module {module_id!r} has no dataset {dataset_id!r}.")
    raw = await loader.run_dataset_route(
        module_id,
        entry.route,
        log_id,
        user_id,
        filter_override=filter_override,
        params=params,
        restrict_event_log=restrict_event_log,
    )
    return adapt(entry.shape, raw)

File: apps/web/lib/visualizations/adapters.ts
/**
 * Client-side `kind → DatasetEnvelope` normalizers (Phase 1).
 *
 * Existing module routes already return `kind`-tagged JSON and already honor
 * the ambient `X-FF-Event-Filter` header + result-cache variant, so a generic
 * viz card just fetches the route and normalizes the response here - no backend
 * change. The adapters are deliberately defensive (introspect arrays into
 * tables, numeric leaves into KPIs) so an unrecognized-but-well-shaped response
 * still renders. Promoted server-side in Phase 2.
 */

import type {
  ColumnSpec,
  ColumnType,
  DatasetEnvelope,
  DatasetShape,
  GraphData,
  GraphEdge,
  GraphNode,
  KpiData,
  KpiMeasure,
  TableData,
  TreeData,
  TreeNode,
} from "./types";

type Json = Record<string, unknown>;

function isObject(v: unknown): v is Json {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function prettify(key: string): string {
  return key
    .replace(/[_-]+/g, " ")
    .replace(/\b\w/g, (c) => c.toUpperCase())
    .trim();
}

const DATE_RE = /^\d{4}-\d{2}-\d{2}([T ]\d{2}:\d{2})?/;

function inferType(value: unknown): ColumnType {
  if (typeof value === "boolean") return "boolean";
  if (typeof value === "number") return Number.isInteger(value) ? "integer" : "number";
  if (typeof value === "string") {
    if (DATE_RE.test(value)) return "datetime";
    const n = Number(value);
    if (value.trim() !== "" && Number.isFinite(n)) return Number.isInteger(n) ? "integer" : "number";
    return "string";
  }
  return "string";
}

function roleFor(type: ColumnType): ColumnSpec["role"] {
  if (type === "number" || type === "integer" || type === "duration") return "measure";
  if (type === "datetime") return "time";
  return "dimension";
}

/** Build a typed table from an array of row objects, inferring columns from the
 * union of keys and the type from each column's first non-null value. */
export function tableFromRows(rows: Json[]): TableData {
  const keys: string[] = [];
  const seen = new Set<string>();
  for (const r of rows) {
    for (const k of Object.keys(r)) {
      if (!seen.has(k)) {
        seen.add(k);
        keys.push(k);
      }
    }
  }
  const columns: ColumnSpec[] = keys.map((k) => {
    const sample = rows.find((r) => r[k] != null)?.[k];
    const type = inferType(sample);
    return { id: k, label: prettify(k), type, role: roleFor(type) };
  });
  return { columns, rows };
}

/** Collect numeric leaves of an object (one level of nesting) into KPI measures. */
function measuresFrom(obj: Json, prefix = ""): KpiMeasure[] {
  const out: KpiMeasure[] = [];
  for (const [k, v] of Object.entries(obj)) {
    const key = prefix ? `${prefix}.${k}` : k;
    if (typeof v === "number") {
      out.push({ key, label: prettify(k), value: v, format: kpiFormat(k, v) });
    } else if (isObject(v)) {
      out.push(...measuresFrom(v, key));
    }
  }
  return out;
}

function kpiFormat(key: string, value: number): KpiMeasure["format"] {
  const k = key.toLowerCase();
  if (/(fitness|precision|perc|percent|share|ratio|rate)/.test(k) && value >= 0 && value <= 1) {
    return "percent";
  }
  if (/(^n[_.]|count|total|num)/.test(k) && Number.isInteger(value)) return "integer";
  return Number.isInteger(value) ? "integer" : "number";
}

function envelope(
  shape: DatasetShape,
  data: DatasetEnvelope["data"],
  columns: ColumnSpec[] = [],
  meta?: DatasetEnvelope["meta"],
): DatasetEnvelope {
  return { shape, schema: { columns }, data, meta };
}

// ── graph ──────────────────────────────────────────────────────────────────

function graphFromDfg(json: Json): GraphData {
  const activities = Array.isArray(json.activities) ? (json.activities as Json[]) : [];
  const edges = Array.isArray(json.edges) ? (json.edges as Json[]) : [];
  const nodes: GraphNode[] = activities.map((a) => ({
    id: String(a.id),
    label: String(a.label ?? a.id),
    value: typeof a.frequency === "number" ? a.frequency : undefined,
    kind: "activity",
  }));
  const gEdges: GraphEdge[] = edges.map((e, i) => ({
    id: String(e.id ?? `e${i}`),
    source: String(e.source),
    target: String(e.target),
    value: typeof e.frequency === "number" ? e.frequency : undefined,
    performanceSeconds: typeof e.performance_seconds === "number" ? e.performance_seconds : undefined,
  }));
  return {
    directed: true,
    nodes,
    edges: gEdges,
    start: Object.keys((json.start_activities as Json) ?? {}),
    end: Object.keys((json.end_activities as Json) ?? {}),
  };
}

function graphFromPetri(json: Json): GraphData {
  const places = Array.isArray(json.places) ? (json.places as Json[]) : [];
  const transitions = Array.isArray(json.transitions) ? (json.transitions as Json[]) : [];
  const arcs = Array.isArray(json.arcs) ? (json.arcs as Json[]) : [];
  const nodes: GraphNode[] = [
    ...places.map((p) => ({ id: String(p.id), label: String(p.label ?? ""), kind: "place" })),
    ...transitions.map((t) => ({
      id: String(t.id),
      label: t.is_invisible ? "" : String(t.label ?? t.name ?? ""),
      kind: "transition",
    })),
  ];
  const edges: GraphEdge[] = arcs.map((a, i) => ({
    id: String(a.id ?? `a${i}`),
    source: String(a.source),
    target: String(a.target),
    value: typeof a.weight === "number" ? a.weight : undefined,
  }));
  return { directed: true, nodes, edges };
}

// ── tree ─────────────────────────────────────────────────────────────────

function treeFromProcessTree(json: Json): TreeData {
  let counter = 0;
  const walk = (node: Json): TreeNode => {
    const children = Array.isArray(node.children) ? (node.children as Json[]) : [];
    const label = (node.label as string) ?? (node.operator as string) ?? "·";
    return {
      id: String(node.id ?? `n${counter++}`),
      label,
      children: children.map(walk),
    };
  };
  const root = isObject(json.root) ? walk(json.root as Json) : { id: "root", label: "·", children: [] };
  return { root };
}

function treeFromPrefix(json: Json): TreeData {
  const flat = Array.isArray(json.nodes) ? (json.nodes as Json[]) : [];
  const byId = new Map<string, TreeNode>();
  for (const n of flat) {
    byId.set(String(n.id), {
      id: String(n.id),
      label: String(n.label ?? "start"),
      value: typeof n.frequency === "number" ? n.frequency : undefined,
      children: [],
    });
  }
  let root: TreeNode | null = null;
  for (const n of flat) {
    const node = byId.get(String(n.id))!;
    const parent = n.parent != null ? byId.get(String(n.parent)) : null;
    if (parent) parent.children.push(node);
    else root = root ?? node;
  }
  return { root: root ?? { id: "root", label: "start", children: [] } };
}

// ── public entry ───────────────────────────────────────────────────────────

/** Normalize a module route's response into the declared shape. `shape` comes
 * from the dataset catalog entry; `json` is the raw route response. */
export function normalize(shape: DatasetShape, json: unknown): DatasetEnvelope {
  const obj: Json = isObject(json) ? json : {};
  const kind = typeof obj.kind === "string" ? obj.kind : "";

  if (shape === "graph") {
    const g = kind === "petri_net" ? graphFromPetri(obj) : graphFromDfg(obj);
    return envelope("graph", g, [], { sourceKind: kind });
  }

  if (shape === "tree") {
    const t = kind === "prefix_tree" ? treeFromPrefix(obj) : treeFromProcessTree(obj);
    return envelope("tree", t, [], { sourceKind: kind });
  }

  if (shape === "kpi") {
    let items: KpiMeasure[] = [];
    let note: string | undefined;
    if (kind === "conformance") {
      if (obj.ran === false || !isObject(obj.kpis)) {
        note = "No conformance run yet - run conformance on this log first.";
      } else {
        items = measuresFrom(obj.kpis as Json);
      }
    } else if (isObject(obj.basic) || isObject(obj.enriched)) {
      // complexity_metrics: basic + enriched numeric leaves.
      items = [
        ...measuresFrom((obj.basic as Json) ?? {}),
        ...(isObject(obj.enriched) ? measuresFrom(obj.enriched as Json, "enriched") : []),
      ];
    } else {
      items = measuresFrom(obj);
    }
    return envelope("kpi", { items } satisfies KpiData, [], { sourceKind: kind, note });
  }

  if (shape === "blob") {
    const value = typeof obj.xml === "string" ? obj.xml : typeof obj.value === "string" ? obj.value : "";
    return envelope("blob", { media: kind === "bpmn" ? "bpmn-xml" : "text", value }, [], { sourceKind: kind });
  }

  // table (default): prefer the known slice, else the first array-of-objects.
  let rows: Json[] = [];
  let note: string | undefined;
  if (kind === "conformance") {
    rows = Array.isArray(obj.per_activity) ? (obj.per_activity as Json[]) : [];
    if (rows.length === 0 && obj.ran === false) note = "No conformance run yet.";
  } else if (Array.isArray(obj.drifts)) {
    rows = obj.drifts as Json[];
  } else {
    for (const v of Object.values(obj)) {
      if (Array.isArray(v) && v.length > 0 && isObject(v[0])) {
        rows = v as Json[];
        break;
      }
    }
  }
  const table = tableFromRows(rows);
  return envelope("table", table, table.columns, {
    sourceKind: kind,
    rowCount: rows.length,
    note: rows.length === 0 ? (note ?? "No rows.") : undefined,
  });
}
